src/log/format.cr

```diff
@@ -0,0 +1,92 @@
+class Log
+  class_property progname = File.basename(PROGRAM_NAME)
+
+  module Formatter
+    abstract def format(entry : Log::Entry, io : IO)
+
+    def self.new(&proc : (Log::Entry, IO) ->)
+      ProcFormatter.new proc
+    end
+  end
+
+  private struct ProcFormatter
+    include Formatter
+
+    def initialize(@proc : (Log::Entry, IO) ->)
+    end
+
+    def format(entry : Log::Entry, io : IO)
+      @proc.call(entry, io)
+    end
+  end
+
+  abstract struct StaticFormat
+    extend Formatter
+
+    def initialize(@entry : Log::Entry, @io : IO)
+    end
+
+    def timestamp
+      @entry.timestamp.to_rfc3339(@io)
+    end
+
+    def string(str)
+      @io << str
+    end
+
+    def message
+      @io << @entry.message
+    end
+
+    def severity
+      @entry.severity.label.rjust(7, @io)
+    end
+
+    def source(*, before = nil, after = nil)
+      if @entry.source.size > 0
+        @io << before << @entry.source << after
+      end
+    end
+
+    def context(*, before = nil, after = nil)
+      if @entry.context.size > 0
+        @io << before << @entry.context << after
+      end
+    end
+
+    def exception(*, before = '\n', after = nil)
+      if ex = @entry.exception
+        @io << before
+        ex.inspect_with_backtrace(@io)
+        @io << after
+      end
+    end
+
+    def progname
+      @io << Log.progname
+    end
+
+    def self.format(entry, io)
+      new(entry, io).run
+    end
+
+    abstract def run
+  end
+
+  macro format(name, pattern)
```

Comment:
For macros that define things, we usually use the `def_` prefix. Or maybe we can name this `define_format`, because `Log.format` seems like it's formatting something, when in reality it's defining a format.

Reply:
Maybe even `define_formatter` because the resulting type is a formatter, not a format.